File: HomeWork1/Hw1/Parser.cs
namespace Hw1;

public static class Parser
{
    public static void ParseCalcArguments(string[] args,
        out double val1,
        out CalculatorOperation operation,
        out double val2)
    {
        if (IsArgLengthSupported(args))
        {
            if (!(double.TryParse(args[0], out val1) && double.TryParse(args[2], out val2)))
                throw new ArgumentException();
            operation = ParseOperation(args[1]);
            if (operation == CalculatorOperation.Undefined) throw new InvalidOperationException();
        }
        else throw new ArgumentException();
    }

    private static bool IsArgLengthSupported(IReadOnlyCollection<string> args) => args.Count == 3;

    private static CalculatorOperation ParseOperation(string arg) => arg switch
    {
        "+" => CalculatorOperation.Plus,
        "-" => CalculatorOperation.Minus,
        "*" => CalculatorOperation.Multiply,
        "/" => CalculatorOperation.Divide,
        _ => CalculatorOperation.Undefined
    };
}

File: Homework4/Hw4/Parser.fs
module Hw4.Parser

open System
open Hw4.CalculatorOperation
open CalcOptions

let isArgLengthSupported (args: string[]) = args.Length = 3

let parseOperation arg =
    match arg with
    | "+" -> CalculatorOperation.Plus
    | "-" -> CalculatorOperation.Minus
    | "*" -> CalculatorOperation.Multiply
    | "/" -> CalculatorOperation.Divide
    | _ -> ArgumentException() |> raise

let parseDouble (str:string) =
    match Double.TryParse str  with
    | true, double -> double
    | _ -> ArgumentException() |> raise
    
let parseCalcArguments args =
   if args = null || not(isArgLengthSupported args) then
       ArgumentException() |> raise
   { arg1 = (parseDouble args[0])
     operation = parseOperation args[1]
     arg2 = (parseDouble args[2]) }